src/audio/dai.c

```diff
@@ -66,6 +67,43 @@ struct dai_data {
 	uint64_t wallclock;	/* wall clock at stream start */
 };
 
+static void dai_atomic_trigger(void *arg, enum notify_id type, void *data);
+
+/* Assign DAI to a group */
+static int dai_assign_group(struct comp_dev *dev, uint32_t group_id)
+{
+	struct dai_data *dd = comp_get_drvdata(dev);
+
+	if (dd->group) {
+		if (dd->group->group_id != group_id) {
+			comp_err(dev, "dai_assign_group(), DAI already in group %d, requested %d",
+				 dd->group->group_id, group_id);
+			return -EINVAL;
+		}
+
+		/* No need to re-assign to the same group, do nothing */
+		return 0;
+	}
+
+	dd->group = dai_group_get(group_id, DAI_CREAT);
+	if (!dd->group) {
+		comp_err(dev, "dai_assign_group(), failed to assign group %d",
+			 group_id);
+		return -EINVAL;
+	}
+
+	dd->group->num_dais++;
```

Comment:
Might be better to encapsulate the counter operations in dai_group_get()/put(). The caller, called by dai_free(), could also clear the group_id setting it to 0 if num_dais goes down to 0. Then the group object could be reused later.